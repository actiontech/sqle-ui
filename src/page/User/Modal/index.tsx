import AddRole from './AddRole';
import AddUser from './AddUser';
import ModifyUserPassword from './ModifyUserPassword';
import UpdateRole from './UpdateRole';
import UpdateUser from './UpdateUser';

const UserManageModal = () => {
  return (
    <>
      <AddRole />
      <UpdateRole />
      <AddUser />
      <UpdateUser />
      <ModifyUserPassword />
    </>
  );
};

export default UserManageModal;
