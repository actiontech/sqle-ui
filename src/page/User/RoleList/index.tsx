import { SyncOutlined } from '@ant-design/icons';
import { useRequest } from 'ahooks';
import {
  Card,
  Button,
  List,
  Typography,
  Space,
  Divider,
  Popconfirm,
  message,
} from 'antd';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { IRoleResV1 } from '../../../api/common';
import role from '../../../api/role';
import EmptyBox from '../../../components/EmptyBox';
import { ResponseCode } from '../../../data/common';
import EmitterKey from '../../../data/EmitterKey';
import { ModalName } from '../../../data/ModalName';
import {
  updateSelectRole,
  updateUserManageModalStatus,
} from '../../../store/userManage';
import EventEmitter from '../../../utils/EventEmitter';
import { RoleListFilter } from './index.type';
import RoleListFilterForm from './RoleListFilterForm';

const RoleList = () => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [roleListFilter, setRoleListFilter] = React.useState<RoleListFilter>(
    {}
  );

  const {
    data,
    loading,
    refresh: refreshRoleList,
    pagination: { total, onChange: changePagination, changeCurrent },
  } = useRequest(
    ({ current, pageSize }) =>
      role.getRoleListV1({
        page_index: current,
        page_size: pageSize,
        ...roleListFilter,
      }),
    {
      paginated: true,
      refreshDeps: [roleListFilter],
      formatResult(res) {
        return {
          list: res.data?.data ?? [],
          total: res.data?.total_nums ?? 0,
        };
      },
    }
  );

  const pageChange = React.useCallback(
    (current: number, pageSize?: number) => {
      if (pageSize) {
        changePagination(current, pageSize);
      } else {
        changeCurrent(current);
      }
    },
    [changeCurrent, changePagination]
  );

  const createRole = () => {
    dispatch(
      updateUserManageModalStatus({
        modalName: ModalName.Add_Role,
        status: true,
      })
    );
  };

  const updateRole = (role: IRoleResV1) => {
    dispatch(updateSelectRole({ role }));
    dispatch(
      updateUserManageModalStatus({
        modalName: ModalName.Update_Role,
        status: true,
      })
    );
  };

  const deleteRole = (roleName: string) => {
    const hideLoading = message.loading(
      t('user.deleteRole.deleting', { name: roleName }),
      0
    );
    role
      .deleteRoleV1({
        role_name: roleName,
      })
      .then((res) => {
        if (res.data.code === ResponseCode.SUCCESS) {
          message.success(
            t('user.deleteRole.deleteSuccessTips', { name: roleName })
          );
          EventEmitter.emit(EmitterKey.Refresh_User_list);
          refreshRoleList();
        }
      })
      .finally(() => {
        setTimeout(() => {
          hideLoading();
        }, 300);
      });
  };

  React.useEffect(() => {
    const refresh = () => {
      refreshRoleList();
    };
    EventEmitter.subscribe(EmitterKey.Refresh_Role_list, refresh);
    return () => {
      EventEmitter.unsubscribe(EmitterKey.Refresh_Role_list, refresh);
    };
  }, [refreshRoleList]);

  return (
    <Card
      title={
        <Space>
          {t('user.roleListTitle')}
          <Button onClick={refreshRoleList}>
            <SyncOutlined spin={loading} />
          </Button>
        </Space>
      }
      extra={[
        <Button key="create-user" type="primary" onClick={createRole}>
          {t('user.createRole.button')}
        </Button>,
      ]}
    >
      <RoleListFilterForm updateRoleListFilter={setRoleListFilter} />
      <List
        loading={loading}
        dataSource={data?.list}
        pagination={{
          total,
          defaultPageSize: 10,
          showSizeChanger: true,
          onChange: pageChange,
        }}
        renderItem={(item) => (
          <List.Item className="user-row-wrapper">
            <List.Item.Meta
              title={item.role_name}
              description={
                item.role_desc || t('user.roleList.roleDescPlaceholder')
              }
            />
            <div className="user-cell">
              <div>{t('user.roleList.username')}</div>
              <EmptyBox
                if={!!item.user_name_list && item.user_name_list.length > 0}
                defaultNode={
                  <Typography.Text disabled>
                    {t('user.roleList.usernamePlaceholder')}
                  </Typography.Text>
                }
              >
                {item.user_name_list?.join(',')}
              </EmptyBox>
            </div>
            <div className="user-cell">
              <div>{t('user.roleList.database')}</div>
              <EmptyBox
                if={
                  !!item.instance_name_list &&
                  item.instance_name_list.length > 0
                }
                defaultNode={
                  <Typography.Text disabled>
                    {t('user.roleList.databasePlaceholder')}
                  </Typography.Text>
                }
              >
                {item.instance_name_list?.join(',')}
              </EmptyBox>
            </div>
            <Space className="user-cell flex-end-horizontal">
              <Typography.Link
                className="pointer"
                onClick={updateRole.bind(null, item)}
              >
                {t('common.edit')}
              </Typography.Link>
              <Divider type="vertical" />
              <Popconfirm
                title={t('user.deleteRole.deleteTips', {
                  name: item.role_name,
                })}
                placement="topRight"
                okText={t('common.ok')}
                cancelText={t('common.cancel')}
                onConfirm={deleteRole.bind(null, item.role_name ?? '')}
              >
                <Typography.Text type="danger" className="pointer">
                  {t('common.delete')}
                </Typography.Text>
              </Popconfirm>
            </Space>
          </List.Item>
        )}
      />
    </Card>
  );
};

export default RoleList;
