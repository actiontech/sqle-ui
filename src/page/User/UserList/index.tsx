import { SyncOutlined } from '@ant-design/icons';
import { useRequest } from 'ahooks';
import {
  message,
  Card,
  Button,
  List,
  Typography,
  Tag,
  Space,
  Divider,
  Popconfirm,
} from 'antd';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { IUserResV1 } from '../../../api/common';
import user from '../../../api/user';
import EmptyBox from '../../../components/EmptyBox';
import { ResponseCode } from '../../../data/common';
import EmitterKey from '../../../data/EmitterKey';
import { ModalName } from '../../../data/ModalName';
import {
  updateSelectUser,
  updateUserManageModalStatus,
} from '../../../store/userManage';
import EventEmitter from '../../../utils/EventEmitter';
import { UserListFilter } from './index.type';
import UserListFilterForm from './UserListFilterForm';

const UserList = () => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [userListFilter, setUserListFilter] = React.useState<UserListFilter>(
    {}
  );

  const {
    data,
    loading,
    refresh: refreshUserList,
    pagination: { total, onChange: changePagination, changeCurrent },
  } = useRequest(
    ({ current, pageSize }) => {
      return user.getUserListV1({
        page_index: current,
        page_size: pageSize,
        ...userListFilter,
      });
    },
    {
      paginated: true,
      refreshDeps: [userListFilter],
      formatResult(res) {
        return {
          list: res.data?.data ?? [],
          total: res.data?.total_nums ?? 0,
        };
      },
    }
  );

  const pageChange = React.useCallback(
    (current: number, pageSize?: number) => {
      if (pageSize) {
        changePagination(current, pageSize);
      } else {
        changeCurrent(current);
      }
    },
    [changeCurrent, changePagination]
  );

  const addUser = React.useCallback(() => {
    dispatch(
      updateUserManageModalStatus({
        modalName: ModalName.Add_User,
        status: true,
      })
    );
  }, [dispatch]);

  const updateUser = React.useCallback(
    (user: IUserResV1) => {
      dispatch(
        updateSelectUser({
          user,
        })
      );
      dispatch(
        updateUserManageModalStatus({
          modalName: ModalName.Update_User,
          status: true,
        })
      );
    },
    [dispatch]
  );

  const removeUser = React.useCallback(
    (username: string) => {
      const hideLoading = message.loading(
        t('user.deleteUser.deleting', { username }),
        0
      );
      user
        .deleteUserV1({ user_name: username })
        .then((res) => {
          if (res.data.code === ResponseCode.SUCCESS) {
            message.success(t('user.deleteUser.deleteSuccess', { username }));
            refreshUserList();
            EventEmitter.emit(EmitterKey.Refresh_Role_list);
          }
        })
        .finally(() => {
          hideLoading();
        });
    },
    [refreshUserList, t]
  );

  React.useEffect(() => {
    const refresh = () => {
      refreshUserList();
    };
    EventEmitter.subscribe(EmitterKey.Refresh_User_list, refresh);
    return () => {
      EventEmitter.unsubscribe(EmitterKey.Refresh_User_list, refresh);
    };
  }, [refreshUserList]);

  return (
    <Card
      title={
        <Space>
          {t('user.userListTitle')}
          <Button onClick={refreshUserList}>
            <SyncOutlined spin={loading} />
          </Button>
        </Space>
      }
      extra={[
        <Button key="create-user" type="primary" onClick={addUser}>
          {t('user.createUser.button')}
        </Button>,
      ]}
    >
      <UserListFilterForm updateUserListFilter={setUserListFilter} />
      <List
        loading={loading}
        dataSource={data?.list}
        pagination={{
          showSizeChanger: true,
          defaultPageSize: 10,
          total,
          onChange: pageChange,
        }}
        renderItem={(item) => (
          <List.Item className="user-row-wrapper">
            <List.Item.Meta
              title={item.user_name}
              description={item.email || t('user.userList.emailPlaceholder')}
            />
            <div className="user-cell">
              <div>{t('user.userList.role')}</div>
              <EmptyBox
                if={!!item.role_name_list && item.role_name_list.length > 0}
                defaultNode={
                  <Typography.Text disabled>
                    {t('user.userList.rolePlaceHolder')}
                  </Typography.Text>
                }
              >
                {item.role_name_list?.map((role) => (
                  <Tag key={role}>{role}</Tag>
                ))}
              </EmptyBox>
            </div>
            <Space className="user-cell flex-end-horizontal">
              <Typography.Link
                className="pointer"
                onClick={updateUser.bind(null, item)}
              >
                {t('common.edit')}
              </Typography.Link>
              <Divider type="vertical" />
              <Popconfirm
                title={t('user.deleteUser.confirmTitle', {
                  username: item.user_name,
                })}
                placement="topRight"
                okText={t('common.ok')}
                cancelText={t('common.cancel')}
                onConfirm={removeUser.bind(null, item.user_name ?? '')}
              >
                <Typography.Text type="danger" className="pointer">
                  {t('common.delete')}
                </Typography.Text>
              </Popconfirm>
            </Space>
          </List.Item>
        )}
      />
    </Card>
  );
};

export default UserList;
