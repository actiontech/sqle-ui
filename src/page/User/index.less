.user-manage-page {
  .user-row-wrapper {
    .user-cell {
      flex: 1;
    }
  }
}
