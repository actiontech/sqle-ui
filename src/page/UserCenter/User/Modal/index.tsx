import AddUser from './AddUser';
import ModifyUserPassword from './ModifyUserPassword';
import UpdateUser from './UpdateUser';

const UserModal = () => {
  return (
    <>
      <AddUser />
      <UpdateUser />
      <ModifyUserPassword />
    </>
  );
};

export default UserModal;
