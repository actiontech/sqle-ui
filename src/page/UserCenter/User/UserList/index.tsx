import { SyncOutlined } from '@ant-design/icons';
import { usePagination } from 'ahooks';
import { message, Card, Button, Space, Table } from 'antd';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { IUserResV1 } from '../../../../api/common';
import user from '../../../../api/user';
import { IGetUserListV1Params } from '../../../../api/user/index.d';
import { ResponseCode } from '../../../../data/common';
import EmitterKey from '../../../../data/EmitterKey';
import { ModalName } from '../../../../data/ModalName';
import {
  updateSelectUser,
  updateUserManageModalStatus,
} from '../../../../store/userManage';
import EventEmitter from '../../../../utils/EventEmitter';
import { UserListFilter } from './index.type';
import tableHeaderFactory from './tableHeader';
import UserListFilterForm from './UserListFilterForm';

const UserList = () => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [userListFilter, setUserListFilter] = React.useState<UserListFilter>(
    {}
  );

  const {
    data,
    loading,
    refresh: refreshUserList,
    pagination: { total, onChange: changePagination, changeCurrent },
  } = usePagination(
    ({ current, pageSize }) => {
      const params: IGetUserListV1Params = {
        page_index: current,
        page_size: pageSize,
        filter_user_name: userListFilter.filter_user_name,
      };
      return user.getUserListV1(params).then((res) => {
        return {
          list: res.data?.data ?? [],
          total: res.data?.total_nums ?? 0,
        };
      });
    },
    {
      refreshDeps: [userListFilter],
    }
  );

  const pageChange = React.useCallback(
    (current: number, pageSize?: number) => {
      if (pageSize) {
        changePagination(current, pageSize);
      } else {
        changeCurrent(current);
      }
    },
    [changeCurrent, changePagination]
  );

  const addUser = React.useCallback(() => {
    dispatch(
      updateUserManageModalStatus({
        modalName: ModalName.Add_User,
        status: true,
      })
    );
  }, [dispatch]);

  const updateUser = React.useCallback(
    (user: IUserResV1) => {
      dispatch(
        updateSelectUser({
          user,
        })
      );
      dispatch(
        updateUserManageModalStatus({
          modalName: ModalName.Update_User,
          status: true,
        })
      );
    },
    [dispatch]
  );

  const updateUserPassword = React.useCallback(
    (user: IUserResV1) => {
      dispatch(
        updateSelectUser({
          user,
        })
      );
      dispatch(
        updateUserManageModalStatus({
          modalName: ModalName.Update_User_Password,
          status: true,
        })
      );
    },
    [dispatch]
  );

  const removeUser = React.useCallback(
    (username: string) => {
      const hideLoading = message.loading(
        t('user.deleteUser.deleting', { username }),
        0
      );
      user
        .deleteUserV1({ user_name: username })
        .then((res) => {
          if (res.data.code === ResponseCode.SUCCESS) {
            message.success(t('user.deleteUser.deleteSuccess', { username }));
            refreshUserList();
            EventEmitter.emit(EmitterKey.Refresh_Role_list);
            EventEmitter.emit(EmitterKey.Refresh_User_Group_List);
          }
        })
        .finally(() => {
          hideLoading();
        });
    },
    [refreshUserList, t]
  );

  React.useEffect(() => {
    const refresh = () => {
      refreshUserList();
    };
    EventEmitter.subscribe(EmitterKey.Refresh_User_list, refresh);
    return () => {
      EventEmitter.unsubscribe(EmitterKey.Refresh_User_list, refresh);
    };
  }, [refreshUserList]);

  return (
    <Card
      title={
        <Space>
          {t('user.userListTitle')}
          <Button onClick={refreshUserList}>
            <SyncOutlined spin={loading} />
          </Button>
        </Space>
      }
      extra={[
        <Button key="create-user" type="primary" onClick={addUser}>
          {t('user.createUser.button')}
        </Button>,
      ]}
    >
      <UserListFilterForm updateUserListFilter={setUserListFilter} />
      <Table
        rowKey="user_name"
        loading={loading}
        dataSource={data?.list}
        columns={tableHeaderFactory(updateUser, removeUser, updateUserPassword)}
        pagination={{
          showSizeChanger: true,
          defaultPageSize: 10,
          total,
          onChange: pageChange,
        }}
      />
    </Card>
  );
};

export default UserList;
