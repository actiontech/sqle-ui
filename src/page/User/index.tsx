import { PageHeader, Space } from 'antd';
import { useTranslation } from 'react-i18next';
import UserManageModal from './Modal';
import RoleList from './RoleList';
import UserList from './UserList';

import './index.less';
import { useDispatch } from 'react-redux';
import { initUserManageModalStatus } from '../../store/userManage';
import { ModalName } from '../../data/ModalName';
import React from 'react';

const User = () => {
  const { t } = useTranslation();
  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(
      initUserManageModalStatus({
        modalStatus: {
          [ModalName.Add_Role]: false,
          [ModalName.Add_User]: false,
          [ModalName.Update_Role]: false,
          [ModalName.Update_User]: false,
          [ModalName.Update_User_Password]: false,
        },
      })
    );
  }, [dispatch]);

  return (
    <article className="user-manage-page">
      <PageHeader title={t('user.pageTitle')} ghost={false}>
        {t('user.pageDesc')}
      </PageHeader>
      <section className="padding-content">
        <Space direction="vertical" className="full-width-element">
          <UserList />
          <RoleList />
        </Space>
      </section>
      <UserManageModal />
    </article>
  );
};

export default User;
