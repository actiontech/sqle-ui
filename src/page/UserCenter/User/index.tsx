import { PageHeader } from 'antd';
import { useTranslation } from 'react-i18next';
import UserModal from './Modal';
import { useDispatch } from 'react-redux';
import { initUserManageModalStatus } from '../../../store/userManage';
import { ModalName } from '../../../data/ModalName';
import React from 'react';
import UserList from './UserList';

const User = () => {
  const { t } = useTranslation();
  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(
      initUserManageModalStatus({
        modalStatus: {
          [ModalName.Add_User]: false,
          [ModalName.Update_User]: false,
          [ModalName.Update_User_Password]: false,
        },
      })
    );
  }, [dispatch]);

  return (
    <article className="user-manage-page">
      <PageHeader title={t('user.pageTitle')} ghost={false}>
        {t('user.pageDesc')}
      </PageHeader>
      <section className="padding-content">
        <UserList />
      </section>
      <UserModal />
    </article>
  );
};

export default User;
