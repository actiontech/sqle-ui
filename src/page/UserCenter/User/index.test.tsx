import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import User from '.';
import { renderWithRedux } from '../../../testUtils/customRender';
import { mockUseDispatch } from '../../../testUtils/mockRedux';

jest.mock('./UserList', () => {
  return () => null;
});

describe('User', () => {
  let dispatchSpy: jest.SpyInstance;

  beforeEach(() => {
    jest.useFakeTimers();
    mockUseDispatch();
    dispatchSpy = mockUseDispatch().scopeDispatch;
  });

  afterEach(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('should match snapshot', async () => {
    const wrapper = shallow(<User />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should init modal status', () => {
    renderWithRedux(<User />);
    expect(dispatchSpy).toBeCalledTimes(1);
    expect(dispatchSpy).toBeCalledWith({
      payload: {
        modalStatus: {
          ADD_USER: false,
          Update_User: false,
          Update_User_Password: false,
        },
      },
      type: 'user/initModalStatus',
    });
  });
});
