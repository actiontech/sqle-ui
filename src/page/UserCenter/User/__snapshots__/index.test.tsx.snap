// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User should match snapshot 1`] = `
<article
  className="user-manage-page"
>
  <PageHeader
    ghost={false}
    title="user.pageTitle"
  >
    user.pageDesc
  </PageHeader>
  <section
    className="padding-content"
  >
    <Component />
  </section>
  <UserModal />
</article>
`;
